### ISSUE TITLE
MacOS High Sierra Apache, Case-insensitive FS causes fork() sanitizer crash.

### ISSUE THREAD

**Issue by jithinraj** [phusion/passenger#1986]
Help please:
https://stackoverflow.com/questions/46628052/phusion-passenger-installation-error-in-macos-10-13-high-sierra

---
```
## Issue report
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: buildout/apache2/module_libboost_oxt.a(regex_debug.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: buildout/apache2/module_libboost_oxt.a(tss_null.o) has no symbols
clang -o buildout/apache2/mod_passenger.o  -Isrc/agent -Isrc/cxx_supportlib -Isrc/cxx_supportlib/vendor-copy -Isrc/cxx_supportlib/vendor-modified -fPIC -D_REENTRANT -I/usr/local/include -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wpointer-arith -Wwrite-strings -Wno-long-long -Wno-missing-field-initializers -fvisibility=hidden -DVISIBILITY_ATTRIBUTE_SUPPORTED -g -DHAS_ALLOCA_H -DHAS_SFENCE -DHAS_LFENCE -DPASSENGER_DEBUG -DBOOST_DISABLE_ASSERTS -c src/apache2_module/mod_passenger.c
src/apache2_module/mod_passenger.c:26:10: fatal error: 'httpd.h' file not found
#include <httpd.h>
         ^~~~~~~~~
1 error generated.
rake aborted!
Command failed with status (1): [clang -o buildout/apache2/mod_passenger.o  -Isrc/agent -Isrc/cxx_supportlib -Isrc/cxx_supportlib/vendor-copy -Isrc/cxx_supportlib/vendor-modified -fPIC -D_REENTRANT -I/usr/local/include -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wpointer-arith -Wwrite-strings -Wno-long-long -Wno-missing-field-initializers -fvisibility=hidden -DVISIBILITY_ATTRIBUTE_SUPPORTED -g -DHAS_ALLOCA_H -DHAS_SFENCE -DHAS_LFENCE -DPASSENGER_DEBUG -DBOOST_DISABLE_ASSERTS -c src/apache2_module/mod_passenger.c]
/private/tmp/passenger-20171008-7392-owxt2w/passenger-5.1.8/build/support/cplusplus.rb:53:in `run_compiler'
/private/tmp/passenger-20171008-7392-owxt2w/passenger-5.1.8/build/support/cplusplus.rb:97:in `compile_c'
/private/tmp/passenger-20171008-7392-owxt2w/passenger-5.1.8/build/support/cplusplus.rb:154:in `block in define_c_object_compilation_task'
Tasks: TOP => apache2 => buildout/apache2/mod_passenger.so => buildout/apache2/mod_passenger.o
```

**Comment by OnixGH** [comment]
High Sierra support was recently [added to stable](https://github.com/phusion/passenger/commit/416894542e3c689f4d609e6ce7f6c9a298ca3992) and will be part of the upcoming release 5.1.9.

**Comment by jithinraj** [comment]
Thanks for the speedy response.

**Comment by patricksereno** [comment]
I don't see that it's fixed in 5.1.11.  I'm running 10.13 (17A405) both with/without "Command Line Tools (macOS High Sierra version 10.13) for Xcode" (version 9.0) installed

I was able to get the compile to complete by modifying apache.rb as such:

	*** apache.rb	2017-10-18 08:33:42.000000000 -0700
	--- apache-fix.rb	2017-10-18 08:34:13.000000000 -0700
	*************** module PhusionPassenger
	*** 614,619 ****
	--- 614,620 ----
	          if os_name_simple == 'macosx' && os_version >= '10.13' && httpd == '/usr/sbin/httpd'
	            xcode_prefix = `/usr/bin/xcode-select -p`.strip
	            flags << "-I#{xcode_prefix}/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/apache2"
	+           flags << "-I/usr/include/apache2"
	          end
	        else
	          apxs2_flags = `#{apxs2} -q CFLAGS`.strip << " -I" << `#{apxs2} -q INCLUDEDIR`.strip
	*************** module PhusionPassenger
	*** 816,821 ****
	--- 817,823 ----
	              xcode_prefix = `/usr/bin/xcode-select -p`.strip
	              ["-I#{xcode_prefix}/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/apr-1",
	                '-lapr-1']
	+             ['-I/usr/include/apr-1', '-lapr-1']
	            else
	              ['-I/usr/include/apr-1', '-lapr-1']
	            end

If you're running home-brew ruby, here's the process

1. Download the above code to /tmp/apache.rb.patch
1. cd /usr/local/lib/ruby/gems/2.4.0/gems/passenger-5.1.11/src/ruby_supportlib/phusion_passenger/platform_info
1. patch < /tmp/apache.rb.patch

**Comment by CamJN** [comment]
I'm on HS and use homebrew to manage Passenger, and it works on my box (I have both xcode and the cli tools installed). Also there is a HS bottle available from homebrew so it works on their CI as well.

Perhaps you could detail the errors you saw?

**Comment by patricksereno** [comment]
This is the log message I’m receiving:

objc[82628]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[82628]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[Wed Oct 18 22:39:47.939850 2017] [core:notice] [pid 53108] AH00052: child pid 82628 exit signal Abort trap (6)

It only happens on the / URI.  If I go to a sub-URI, such as /apple-touch-icon-152x152-precomposed.png, the service works fine.

I’m going to try the bottle and I’ll let you know if this fixes it.

Best regards,
Patrick Sereno

**Comment by FooBarWidget** [comment]
@ patricksereno It looks like your High Sierra system is a bit different than the ones we tested on. Can you tell me what you see if you run this command?

    find / -name httpd.h

**Comment by patricksereno** [comment]
/usr/include/apache2/httpd.h
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/apache2/httpd.h

Patrick Sereno

**Comment by FooBarWidget** [comment]
Thanks @patricksereno. I've published a branch `GH-1986-high-sierra-continued-fix`. Can you check whether that one compiles correctly?

And can you also show me what `sw_vers` outputs?

@CamJN As for his `__NSCFConstantString initialize` error: the message indicates that it's an Apache worker process that's crashing. Are you able to reproduce this on your system with Apache?

**Comment by patricksereno** [comment]
This is the log dump from running passenger from the home-brew bottle:

[ N 2017-10-18 23:06:42.2085 93306/T1 age/Wat/WatchdogMain.cpp:1267 ]: Starting Passenger watchdog...
[ N 2017-10-18 23:06:42.3248 93308/T1 age/Cor/CoreMain.cpp:1165 ]: Starting Passenger core...
[ N 2017-10-18 23:06:42.3260 93308/T1 age/Cor/CoreMain.cpp:249 ]: Passenger core running in multi-application mode.
[ W 2017-10-18 23:06:42.3443 93308/T1 age/Cor/CoreMain.cpp:897 ]: WARNING: potential privilege escalation vulnerability. Passenger is running as root, and part(s) of the passenger root path (/usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger/locations.ini) can be changed by non-root user(s):
The path "/usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger/locations.ini" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec/src/ruby_supportlib" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec/src" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
[ N 2017-10-18 23:06:42.3444 93308/T1 age/Cor/CoreMain.cpp:905 ]: Passenger core online, PID 93308

[ N 2017-10-18 23:06:42.4692 93308/T7 age/Cor/CoreMain.cpp:589 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
[ N 2017-10-18 23:06:42.4694 93308/T1 age/Cor/CoreMain.cpp:1080 ]: Received command to shutdown gracefully. Waiting until all clients have disconnected...
[ N 2017-10-18 23:06:42.4701 93308/Tn Ser/Server.h:885 ]: [ApiServer] Freed 0 spare client objects
[ N 2017-10-18 23:06:42.4702 93308/T9 Ser/Server.h:885 ]: [ServerThr.2] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4703 93308/Tb Ser/Server.h:885 ]: [ServerThr.3] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4703 93308/T9 Ser/Server.h:531 ]: [ServerThr.2] Shutdown finished
[ N 2017-10-18 23:06:42.4704 93308/Tb Ser/Server.h:531 ]: [ServerThr.3] Shutdown finished
[ N 2017-10-18 23:06:42.4704 93308/Tg Ser/Server.h:885 ]: [ServerThr.5] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4704 93308/Th Ser/Server.h:885 ]: [ServerThr.6] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4703 93308/T7 Ser/Server.h:885 ]: [ServerThr.1] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4703 93308/Tn Ser/Server.h:531 ]: [ApiServer] Shutdown finished
[ N 2017-10-18 23:06:42.4705 93308/Td Ser/Server.h:885 ]: [ServerThr.4] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4706 93308/Tl Ser/Server.h:885 ]: [ServerThr.8] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4705 93308/Tk Ser/Server.h:885 ]: [ServerThr.7] Freed 128 spare client objects
[ N 2017-10-18 23:06:42.4708 93308/Th Ser/Server.h:531 ]: [ServerThr.6] Shutdown finished
[ N 2017-10-18 23:06:42.4709 93308/T7 Ser/Server.h:531 ]: [ServerThr.1] Shutdown finished
[ N 2017-10-18 23:06:42.4710 93308/Td Ser/Server.h:531 ]: [ServerThr.4] Shutdown finished
[ N 2017-10-18 23:06:42.4711 93308/Tl Ser/Server.h:531 ]: [ServerThr.8] Shutdown finished
[ N 2017-10-18 23:06:42.4707 93308/Tg Ser/Server.h:531 ]: [ServerThr.5] Shutdown finished
[ N 2017-10-18 23:06:42.4711 93308/Tk Ser/Server.h:531 ]: [ServerThr.7] Shutdown finished
[ N 2017-10-18 23:06:42.5302 93311/T1 age/Wat/WatchdogMain.cpp:1267 ]: Starting Passenger watchdog...
[ N 2017-10-18 23:06:42.5641 93313/T1 age/Cor/CoreMain.cpp:1165 ]: Starting Passenger core...
[ N 2017-10-18 23:06:42.5653 93313/T1 age/Cor/CoreMain.cpp:249 ]: Passenger core running in multi-application mode.
[ W 2017-10-18 23:06:42.5843 93313/T1 age/Cor/CoreMain.cpp:897 ]: WARNING: potential privilege escalation vulnerability. Passenger is running as root, and part(s) of the passenger root path (/usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger/locations.ini) can be changed by non-root user(s):
The path "/usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger/locations.ini" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec/src/ruby_supportlib" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec/src" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger/libexec" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt/passenger" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
The path "/usr/local/opt" can be modified by user "administrator" (or applications running as that user). Change the owner of the path to root, or avoid running Passenger as root.
[ N 2017-10-18 23:06:42.5844 93313/T1 age/Cor/CoreMain.cpp:905 ]: Passenger core online, PID 93313
[Wed Oct 18 23:06:42.587272 2017] [mpm_prefork:notice] [pid 93301] AH00163: Apache/2.4.27 (Unix) Phusion_Passenger/5.1.11 LibreSSL/2.2.7 configured -- resuming normal operations
[Wed Oct 18 23:06:42.587344 2017] [core:notice] [pid 93301] AH00094: Command line: '/usr/sbin/httpd -D SERVER_APP_HAS_DEFAULT_PORTS -D FOREGROUND -f /Library/Server/Web/Config/apache2/httpd_server_app.conf -E /var/log/apache2/error_log -D WEBSERVICE_ON'
[ N 2017-10-18 23:06:42.9844 93308/T1 age/Cor/CoreMain.cpp:1150 ]: Passenger core shutdown finished
[ N 2017-10-18 23:06:46.0801 93313/T5 age/Cor/SecurityUpdateChecker.h:374 ]: Security update check: no update found (next check in 24 hours)
objc[93521]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[93521]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[Wed Oct 18 23:08:28.389155 2017] [core:notice] [pid 93301] AH00052: child pid 93521 exit signal Abort trap (6)


This actually made things worse, now
Patrick Sereno
(602) 723-3166
patrick@6orbits.com

**Comment by CamJN** [comment]
@patricksereno Are you running Apache with the event, worker, or prefork MPM module in use?

**Comment by patricksereno** [comment]
prefork.  I'm using macOS Server Version 5.4 (17S1207).

This appears to default to prefork.

    httpd -D DUMP_MODULES | grep mpm
     mpm_prefork_module (static)

**Comment by CamJN** [comment]
@patricksereno could you apply the following patch and recompile? There will be some clang warnings about unused flags, that's fine. I haven't been able to reproduce your issue, unfortunately so I can't verify it myself.
```
diff --git a/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb b/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb
index e546f0798..f4dbf9d35 100644
--- a/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb
+++ b/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb
@@ -593,7 +593,7 @@ def self.apache2_module_cxxflags(with_apr_flags = true)
     # The C compiler flags that are necessary to compile an Apache module.
     # Also includes APR and APU compiler flags if with_apr_flags is true.
     def self.apache2_module_c_or_cxxflags(language, with_apr_flags = true)
-      flags = [""]
+      flags = ["-framework", "CoreFoundation"]
       if (language == :c && cc_is_sun_studio?) || (language == :cxx && cxx_is_sun_studio?)
         flags << "-KPIC"
       else
```

**Comment by CamJN** [comment]
@patricksereno could you apply the following patch and recompile? There will be some clang warnings about unused flags, that's fine. I haven't been able to reproduce your issue, unfortunately so I can't verify it myself.
```
diff --git a/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb b/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb
index e546f0798..fe075c21c 100644
--- a/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb
+++ b/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb
@@ -593,7 +593,7 @@ def self.apache2_module_cxxflags(with_apr_flags = true)
     # The C compiler flags that are necessary to compile an Apache module.
     # Also includes APR and APU compiler flags if with_apr_flags is true.
     def self.apache2_module_c_or_cxxflags(language, with_apr_flags = true)
-      flags = [""]
+      flags = ["-framework", "Foundation"]
       if (language == :c && cc_is_sun_studio?) || (language == :cxx && cxx_is_sun_studio?)
         flags << "-KPIC"
       else
```

**Comment by patricksereno** [comment]
Below is the output from attempted compile/install:

$ bin/passenger-install-apache2-module
Welcome to the Phusion Passenger Apache 2 module installer, v5.1.12.

This installer will guide you through the entire installation process. It
shouldn't take more than 3 minutes in total.

Here's what you can expect from the installation process:

 1. The Apache 2 module will be installed for you.
 2. You'll learn how to configure Apache.
 3. You'll learn how to deploy a Ruby on Rails application.

Don't worry if anything goes wrong. This installer will advise you on how to
solve any problems.

Press Enter to continue, or Ctrl-C to abort.


--------------------------------------------

Which languages are you interested in?

Use <space> to select.
If the menu doesn't display correctly, press '!'

   ⬢  Ruby
   ⬡  Python
 ‣ ⬡  Node.js
   ⬡  Meteor

--------------------------------------------

Checking for required software...

 * Checking for C compiler...
      Found: yes
      Location: /usr/bin/cc
 * Checking for C++ compiler...
      Found: yes
      Location: /usr/bin/c++
 * Checking for Curl development headers with SSL support...
      Found: yes
      curl-config location: /usr/bin/curl-config
      Header location: /usr/include/curl/curl.h
      Version: libcurl 7.54.0
      Usable: yes
      Supports SSL: yes
 * Checking for Zlib development headers...
      Found: yes
      Location: /usr/include/zlib.h
 * Checking for Apache 2...
      Found: yes
      Location of httpd: /usr/sbin/httpd
      Apache version: 2.4.27
 * Checking for Rake (associated with /usr/local/Cellar/ruby/2.4.2/bin/ruby)...
      Found: yes
      Location: /usr/local/Cellar/ruby/2.4.2/bin/ruby /usr/local/Cellar/ruby/2.4.2/bin/rake
 * Checking for OpenSSL support for Ruby...
      Found: yes
 * Checking for RubyGems...
      Found: yes
 * Checking for Ruby development headers...
      Found: yes
      Location: /usr/local/Cellar/ruby/2.4.2/include/ruby-2.4.0/ruby.h
 * Checking for rack...
      Found: yes

--------------------------------------------

Checking whether there are multiple Apache installations...
Only a single installation detected. This is good.

--------------------------------------------
Compiling and installing Apache 2 module...
cd /Users/administrator/passenger
# env NOEXEC_DISABLE=1 /usr/local/Cellar/ruby/2.4.2/bin/ruby /usr/local/Cellar/ruby/2.4.2/bin/rake RELEASE=yes apache2:clean apache2
rm -rf buildout/cache
rm -rf buildout/common/libboost_oxt.a buildout/common/libboost_oxt
rm -f buildout/common/libpassenger_common/LoggingKit.o buildout/common/libpassenger_common/Exceptions.o buildout/common/libpassenger_common/ProcessManagement/Spawn.o buildout/common/libpassenger_common/ProcessManagement/Utils.o buildout/common/libpassenger_common/Utils/SystemTime.o buildout/common/libpassenger_common/Utils/StrIntUtils.o buildout/common/libpassenger_common/Utils/StrIntUtilsNoStrictAliasing.o buildout/common/libpassenger_common/Utils/IOUtils.o buildout/common/libpassenger_common/Utils/Hasher.o buildout/common/libpassenger_common/Utils.o buildout/common/libpassenger_common/jsoncpp.o
rm -f buildout/common/libpassenger_common/Crypto.o buildout/common/libpassenger_common/Utils/CachedFileStat.o buildout/common/libpassenger_common/Utils/LargeFiles.o buildout/common/libpassenger_common/WatchdogLauncher.o buildout/common/libpassenger_common/MemoryKit/mbuf.o buildout/common/libpassenger_common/MemoryKit/palloc.o buildout/common/libpassenger_common/ServerKit/http_parser.o buildout/common/libpassenger_common/ServerKit/Implementation.o buildout/common/libpassenger_common/DataStructures/LString.o buildout/common/libpassenger_common/AppTypes.o
rm -f buildout/common/libpassenger_common/vendor-modified/modp_b64.o buildout/common/libpassenger_common/vendor-modified/modp_b64_strict_aliasing.o
rm -f buildout/common/libpassenger_common/UnionStationFilterSupport.o
rm -f buildout/common/libpassenger_common/ProcessManagement/Ruby.o
rm -rf buildout/common/libpassenger_common
rm -rf buildout/support-binaries/
rm -rf buildout/apache2/module_libboost_oxt.a buildout/apache2/module_libboost_oxt
rm -f buildout/apache2/module_libpassenger_common/LoggingKit.o buildout/apache2/module_libpassenger_common/Exceptions.o buildout/apache2/module_libpassenger_common/ProcessManagement/Spawn.o buildout/apache2/module_libpassenger_common/ProcessManagement/Utils.o buildout/apache2/module_libpassenger_common/Utils/SystemTime.o buildout/apache2/module_libpassenger_common/Utils/StrIntUtils.o buildout/apache2/module_libpassenger_common/Utils/StrIntUtilsNoStrictAliasing.o buildout/apache2/module_libpassenger_common/Utils/IOUtils.o buildout/apache2/module_libpassenger_common/Utils/Hasher.o buildout/apache2/module_libpassenger_common/Utils.o buildout/apache2/module_libpassenger_common/jsoncpp.o
rm -f buildout/apache2/module_libpassenger_common/Crypto.o buildout/apache2/module_libpassenger_common/Utils/CachedFileStat.o buildout/apache2/module_libpassenger_common/Utils/LargeFiles.o buildout/apache2/module_libpassenger_common/WatchdogLauncher.o buildout/apache2/module_libpassenger_common/MemoryKit/mbuf.o buildout/apache2/module_libpassenger_common/MemoryKit/palloc.o buildout/apache2/module_libpassenger_common/ServerKit/http_parser.o buildout/apache2/module_libpassenger_common/ServerKit/Implementation.o buildout/apache2/module_libpassenger_common/DataStructures/LString.o buildout/apache2/module_libpassenger_common/AppTypes.o
rm -f buildout/apache2/module_libpassenger_common/vendor-modified/modp_b64.o buildout/apache2/module_libpassenger_common/vendor-modified/modp_b64_strict_aliasing.o
rm -f buildout/apache2/module_libpassenger_common/UnionStationFilterSupport.o
rm -f buildout/apache2/module_libpassenger_common/ProcessManagement/Ruby.o
rm -rf buildout/apache2/module_libpassenger_common
rm -rf buildout/apache2/mod_passenger.o buildout/apache2/Configuration.o buildout/apache2/Bucket.o buildout/apache2/Hooks.o buildout/apache2/mod_passenger.so
rake aborted!
LoadError: cannot load such file -- /Users/administrator/passenger/build/support/vendor/cxxcodebuilder/lib/cxxcodebuilder
/Users/administrator/passenger/Rakefile:42:in `require_build_system_file'
/Users/administrator/passenger/build/support/general.rb:28:in `initialize'
/Users/administrator/passenger/build/common_library.rb:259:in `new'
/Users/administrator/passenger/build/common_library.rb:259:in `block in <top (required)>'
/usr/local/lib/ruby/gems/2.4.0/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
Tasks: TOP => apache2 => buildout/apache2/mod_passenger.so => buildout/apache2/module_libpassenger_common/LoggingKit.o => src/cxx_supportlib/Constants.h
(See full trace by running task with --trace)

--------------------------------------------

It looks like something went wrong

Please read our documentation for troubleshooting tips:

   https://www.phusionpassenger.com/library/install/apache/
   https://www.phusionpassenger.com/library/admin/apache/troubleshooting/

If that doesn't help, please use our support facilities. We'll do our best to help you.

   https://www.phusionpassenger.com/support

$ git branch
* GH-1986-high-sierra-continued-fix
  stable-5.1

**Comment by CamJN** [comment]
@patricksereno can you try and run just `rake apache2` in the passenger repo, there's occasionally a problem with running `rake apache2:clean apache2` in one step, which is what `passenger-install-apache2-module` does behind the scenes.

**Comment by wulffeld** [comment]
I also have this issue. Oddly on two different Macs one works just fine with 99.9% the same Apache2 config.

I tried the GH-1986-high-sierra-continued-fix branch as well to no avail. Let me know if I can assist in tracking this down. :)

**Comment by CamJN** [comment]
@wulffeld which part of the issue? The compilation problem or the forking issue?

**Comment by wulffeld** [comment]
@CamJN Ah sorry. The fork issue.

**Comment by CamJN** [comment]
@wulffeld Are the same apps being run on each mac?

**Comment by wulffeld** [comment]
@CamJN Yes, same Rails apps.

**Comment by CamJN** [comment]
@wulffeld Ok what installation method do you use? If you use a package manager are you compiling locally? Does your affected box officially support High Sierra or is it patched to allow the install? Is there anything else you can think of that would differentiate your box from another?

**Comment by wulffeld** [comment]
My box that fails is "MacBook Pro (Retina, 15-inch, Late 2013)" so it supports High Sierra out of the box.

I've tried:
  - homebrew
  - regular *gem install passenger* as a system gem
  - installed from the branch mentioned above.

Both /etc/apache2 dirs are for all intents and purposes the same.

**Comment by CamJN** [comment]
@wulffeld just out of curiosity what does `openssl sha256 < $(which httpd)` output?

**Comment by wulffeld** [comment]
@CamJN f0c361a35a3afd747f628991733ed8d0f1a5c098217311bacbae13f31719d7cc - same on the good mac.

Both have the same Xcode and cli tools as far as I can tell.

**Comment by wulffeld** [comment]
@CamJN So sorry, but the passenger that works was actually a version 5.1.2. Now what I did was rsync /usr/local/Cellar/passenger/5.1.2 to the failing box, do brew switch passenger 5.1.2, and the application(s) work just fine. (I couldn't easily figure out how to install and old version via brew).

So for now I'm good but I can now confirm that 5.1.11 crashes with the same fork() error on both boxes which perhaps makes more sense. :-/

**Comment by CamJN** [comment]
@wulffeld well it certainly makes more sense than it only happening on one box. Though why it doesn't happen on my box is still weird. I've got the homebrew bottled passenger and system apache in use, and don't see the crash.

Did you perhaps not update to apfs?

**Comment by wulffeld** [comment]
@CamJN The volume on one box is apfs. On the other Extended (Journaled).

**Comment by OnixGH** [comment]
@wulffeld and can you tell us which one has which volume? :)

**Comment by wulffeld** [comment]
@OnixGH ? My new Mac has Extended, the old 2013 has apfs. They both fail though so I doubt that's the issue :)

**Comment by NDuggan** [comment]
I have the same fork() error on High Sierra with Passenger 5.1.11 😞

**Comment by CamJN** [comment]
@wulffeld & @NDuggan can you both run `httpd -M` and post the results?

**Comment by NDuggan** [comment]
Hi @CamJN,

Thanks for the quick reply.

**Output**
```
$ httpd -M
AH00112: Warning: DocumentRoot [/Users/[REDACTED]/Code/rails_app/public] does not exist
AH00557: httpd: apr_sockaddr_info_get() failed for Nialls-MacBook-Pro-2.local
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 127.0.0.1. Set the 'ServerName' directive globally to suppress this message
Loaded Modules:
 core_module (static)
 so_module (static)
 http_module (static)
 mpm_prefork_module (static)
 authn_file_module (shared)
 authn_core_module (shared)
 authz_host_module (shared)
 authz_groupfile_module (shared)
 authz_user_module (shared)
 authz_core_module (shared)
 access_compat_module (shared)
 auth_basic_module (shared)
 reqtimeout_module (shared)
 filter_module (shared)
 mime_module (shared)
 log_config_module (shared)
 env_module (shared)
 headers_module (shared)
 setenvif_module (shared)
 version_module (shared)
 slotmem_shm_module (shared)
 unixd_module (shared)
 status_module (shared)
 autoindex_module (shared)
 negotiation_module (shared)
 dir_module (shared)
 alias_module (shared)
 hfs_apple_module (shared)
 passenger_module (shared)
```

**Comment by CamJN** [comment]
@NDuggan this is probably not goign to do anything about the fork issue, but it would be a good idea to fix the DocumentRoot in your virtualhost to point to a real dir, even if it's empty.

**Comment by NDuggan** [comment]
@CamJN Correct. I removed it after I submitted the above output.

**Comment by wulffeld** [comment]
@CamJN 

```
Loaded Modules:
 core_module (static)
 so_module (static)
 http_module (static)
 mpm_prefork_module (static)
 authn_file_module (shared)
 authn_core_module (shared)
 authz_host_module (shared)
 authz_groupfile_module (shared)
 authz_user_module (shared)
 authz_core_module (shared)
 access_compat_module (shared)
 auth_basic_module (shared)
 reqtimeout_module (shared)
 filter_module (shared)
 deflate_module (shared)
 mime_module (shared)
 log_config_module (shared)
 env_module (shared)
 headers_module (shared)
 setenvif_module (shared)
 version_module (shared)
 proxy_module (shared)
 slotmem_shm_module (shared)
 unixd_module (shared)
 status_module (shared)
 autoindex_module (shared)
 negotiation_module (shared)
 dir_module (shared)
 alias_module (shared)
 hfs_apple_module (shared)
 passenger_module (shared)
```

**Comment by CamJN** [comment]
Well, narrowing my modules down to just the common subset doesn't repro the issue. Is everyone using the smart spawn method? And is everyone using apachectl or launchctl to manage apache's lifecycle?

**Comment by NDuggan** [comment]
@CamJN I don't know if I'm using the smart spawn method? But I am using `apachectl` to manage the lifecycle.

**Comment by wulffeld** [comment]
@CamJN Using apachectl and using smart spawning. Disabling smart spawning didn't change anything.

@NDuggan See https://www.phusionpassenger.com/library/indepth/ruby/spawn_methods/.

**Comment by NDuggan** [comment]
@wulffeld thank you.

I must be using the same then, as I'm only using it for Ruby apps and smart spawning appears to be the default 👍

**Comment by CamJN** [comment]
Can either of you reproduce with a default rails app (with postgres db to ensure obj-c gets loaded, sqlite probably won't trigger this bug)? I'm wondering if there might be a badly behaved gem out there...

**Comment by wulffeld** [comment]
@CamJN I have a few sample projects lying around. One old Rails 5.0.1 test app - basicly 100% clean also fails. The Gemfile:

```
source 'https://rubygems.org'

git_source(:github) do |repo_name|
  repo_name = "#{repo_name}/#{repo_name}" unless repo_name.include?("/")
  "https://github.com/#{repo_name}.git"
end


# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
gem 'rails', '~> 5.0.1'
# Use sqlite3 as the database for Active Record
gem 'sqlite3'
# Use Puma as the app server
gem 'puma', '~> 3.0'
# Use SCSS for stylesheets
gem 'sass-rails', '~> 5.0'
# Use Uglifier as compressor for JavaScript assets
gem 'uglifier', '>= 1.3.0'
# Use CoffeeScript for .coffee assets and views
gem 'coffee-rails', '~> 4.2'
# See https://github.com/rails/execjs#readme for more supported runtimes
# gem 'therubyracer', platforms: :ruby

# Use jquery as the JavaScript library
gem 'jquery-rails'
# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks
# gem 'turbolinks', '~> 5'
# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
gem 'jbuilder', '~> 2.5'
# Use Redis adapter to run Action Cable in production
# gem 'redis', '~> 3.0'
# Use ActiveModel has_secure_password
# gem 'bcrypt', '~> 3.1.7'

# Use Capistrano for deployment
# gem 'capistrano-rails', group: :development

group :development, :test do
  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
  gem 'byebug', platform: :mri
end

group :development do
  # Access an IRB console on exception pages or by using <%= console %> anywhere in the code.
  gem 'web-console', '>= 3.3.0'
  gem 'listen', '~> 3.0.5'
  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
  gem 'spring'
  gem 'spring-watcher-listen', '~> 2.0.0'
end

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
gem 'pg'
```

**Comment by wulffeld** [comment]
@NDuggan Interesting find with the sub paths. Same experience here. Root and 404 URLs fails.

**Comment by CamJN** [comment]
@wulffeld does the issue still occur if you remove/comment-out puma from the gemfile?

Can you guys set the PassengerLogLevel to 7 https://www.phusionpassenger.com/library/config/apache/reference/#passengerloglevel and capture the logs that occur when a request to the root path fails?

**Comment by CamJN** [comment]
@wulffeld are you also using the `.dev` tld, and if so what method are you using to redirect those requests to localhost? @NDuggan same question about the DNS.

**Comment by CamJN** [comment]
This is my virtualhost for my test app which I've configured to use the same gemfile as above:
```
<VirtualHost *:80>
    ServerAdmin webmaster@memory.dev
    DocumentRoot "/Users/camdennarzt/Sites/memory/public"
    ServerName memory.dev
    ServerAlias www.memory.dev
    ErrorLog "/private/var/log/apache2/memory.dev-error_log"
    CustomLog "/private/var/log/apache2/memory.dev-access_log" common

    PassengerAppEnv development
    PassengerAppRoot "/Users/camdennarzt/Sites/memory"
    <Directory "/Users/camdennarzt/Sites/memory/public">
        Options +FollowSymLinks -MultiViews -Indexes
        AllowOverride All
        Require all granted
    </Directory>
</VirtualHost>
```
Is there anything that stands out as being different than your setups?

**Comment by wulffeld** [comment]
@CamJN https://gist.github.com/wulffeld/03e7d85caaf05244ecc3d6af70537f4e - at 18:45:50 is where I try to load root. It's like passenger doesn't even get to write to the logs. A request that succeeds has a lot more info (is not included in the gist).

I'm using my own DNS server which resolves it to 127.0.0.1. My app would be for instance 'scr.dev.lan'. No change if I hardcode that in /etc/hosts to 127.0.0.1.

My vhost config would be:

```
<VirtualHost *:80>
  ServerName scr.dev.lan
  DocumentRoot "/Users/m/projects/rails/rails5test"

  RackEnv development

  <directory "/Users/m/projects/rails/rails5test">
    Require all granted
  </directory>
</VirtualHost>
```

No change if I make modifications so it looks like your example.

**Comment by CamJN** [comment]
@wulffeld what does the access log say? It's very strange that Passenger doesn't log anything related to the request at that log level. As you saw with the other request there should be a tonne of output.

Does curl have different results than chrome/safari? It might be a caching issue if so.

**Comment by wulffeld** [comment]
@CamJN The access_log doesn't record the failing attempts.

error_log is full of:

```
objc[51478]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[51478]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[Mon Nov 13 19:08:23.451913 2017] [core:notice] [pid 51413] AH00052: child pid 51478 exit signal Abort trap (6)
```

curl absolutely the same:

```
$ curl http://scr.dev.lan/
curl: (52) Empty reply from server
```

(and same result in error log)

**Comment by CamJN** [comment]
@wulffeld If the access log doesn't record the failing attempts that actually suggests that this might be an apache issue not a passenger issue. Could you setup a static html virtualhost and test that real quick?

**Comment by wulffeld** [comment]
@CamJN You are right. A simple static site produces the same fork() error but works fine with 5.1.2. The only passenger related conf is in /etc/apache/other/passenger.conf:

The static vhost fails even with *PassengerEnabled off* added.

```
LoadModule passenger_module /usr/local/opt/passenger/libexec/buildout/apache2/mod_passenger.so
PassengerRoot /usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger/locations.ini
PassengerDefaultRuby /Users/m/.rbenv/versions/2.4.1/bin/ruby

PassengerLogLevel 7
PassengerLogFile /usr/local/var/log/passenger.log

PassengerDisableSecurityUpdateCheck on
```

**Comment by CamJN** [comment]
@wulffeld can you try the `GH-1986-high-sierra-continued-fix` branch again?

**Comment by wulffeld** [comment]
@CamJN No dice, same error. I pulled, cleaned and did the ./bin/passenger-install-apache2-module dance.

**Comment by CamJN** [comment]
Hate to be pedantic but were you sure to update the `LoadModule` line in your apache config? I know I forget to update that sometimes.

**Comment by wulffeld** [comment]
@CamJN No problem. :) Yeah, I did.

**Comment by wulffeld** [comment]
@CamJN Very curious why there's a difference between root and a sub path in passengers start up process. Wouldn't the error lie there somewhere?

**Comment by NDuggan** [comment]
@CamJN yes, I'm using the .dev tld and directing it via `dnsmasq` - likewise with @wulffeld a manual entry in the hosts file makes no difference. Also if I access `0.0.0.0` or `127.0.0.1` I get the same "empty reply from server" but I can access subpaths via the IPs.

I'm using `unicorn` instead of `puma` but tried changing it and commenting them all out and it made no difference.

Curl shows the same thing as Chrome/Safari + this is on a new MBP with no previous cache of the apps.

Same error for a fresh rails app with no logs going into `logs/development.log` or `/var/log/apache2/access_log` but the same error as above in `/var/log/apache2/error_log`.

**Comment by CamJN** [comment]
@NDuggan if you are using Passenger you don't need to use puma or unicorn or any other app server, as that's what Passenger does.

@wulffeld yeah the root versus subpath thing is indeed the best lead we've got so far, though I'm not sure exactly what it implies yet. The fact that you were able to reproduce with Passenger disabled implies it's simply enough for mod_passenger to be loaded and for apache to process a root-path request.

**Comment by CamJN** [comment]
Do either of you have SIP disabled?

**Comment by wulffeld** [comment]
@CamJN Sorry, what's SIP?

**Comment by CamJN** [comment]
System integrity protection, also known as rootless. It’s a security measure in macOS as of El Capitan. It prevents you modifying the OS in various ways, even as an admin. It’s on by default on most macs.

Sent from my iPhone

On Nov 13, 2017, at 2:38 PM, Martin Moen Wulffeld <notifications@github.com<mailto:notifications@github.com>> wrote:


@CamJN<https://github.com/camjn> Sorry, what's SIP?

**Comment by wulffeld** [comment]
@CamJN 

```$ csrutil status
System Integrity Protection status: enabled.
```

**Comment by NDuggan** [comment]
@CamJN Same here: `System Integrity Protection status: enabled.`

**Comment by NDuggan** [comment]
@CamJN @wulffeld I have a fix - which might prove it's an apache bug rather than passenger.

I basically replaced the system apache with homebrew's apache.

```
# stop system apache
sudo apachectl stop
sudo launchctl unload -w /System/Library/LaunchDaemons/org.apache.httpd.plist 2>/dev/null

# install with homebrew (http.conf located in /usr/local/etc/httpd/http.conf)
brew install httpd
sudo brew services start httpd
```

Note: This brew service must be started with sudo, otherwise it won't work (even though it will say it started). Alternatively you can ensure it's in your path, open a terminal or reload source and then`sudo apachectl start`.

Hope this helps.

**Comment by OnixGH** [comment]
@NDuggan did you verify that your homebrew Apache also loads the Passenger module? (i.e. it has the LoadModule configuration for Passenger, and it recognizes Passenger directives like `PassengerEnabled off`)

**Comment by NDuggan** [comment]
@OnixGH I think so to the first part. I didn't try `PassengerEnabled off` though…

I created this file inside the homebrew apache dir (`/usr/local/etc/httpd/other/passenger.conf`):
```
LoadModule passenger_module /usr/local/opt/passenger/libexec/buildout/apache2/mod_passenger.so
PassengerRoot /usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger/locations.ini
PassengerDefaultRuby /Users/niall/.rbenv/shims/ruby
PassengerAppEnv development
```

And then included this in the homebrew apache httpd.conf (`/usr/local/etc/httpd/httpd.conf`):
```
Include /usr/local/etc/httpd/other/*.conf

<Directory "/Users/niall/Code">
    Options FollowSymLinks Multiviews
    MultiviewsMatch Any
    AllowOverride None
    Require all granted
</Directory>

PassengerAppEnv development

<VirtualHost *:80>
  DocumentRoot "/Users/niall/Code/rails-app/public"
  ServerName rails-app.dev
</VirtualHost>
```

**Comment by OnixGH** [comment]
@NDuggan ah, no that's fine, I thought you were also testing with a static site, but actually using Passenger is an even better test :)

**Comment by CamJN** [comment]
Ok, in order to get to the bottom of the system apache issue, here is what I believe is required.

On an affected system, one would have to enable core dumps, and disable SIP. Then either provide us with the core dump file or allow myself to join you in a skype screensharing session so we can investigate the problem.

The procedure for enabling core dumps is below:
```
echo '<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>core</key>
    <array>
      <integer>-1</integer>
      <integer>-1</integer>
    </array>
  </dict>
</plist>' | sudo tee -a /com.apple.xpc.launchd.limits.plist
sudo chown root:wheel /com.apple.xpc.launchd.limits.plist
sudo chmod 644 /com.apple.xpc.launchd.limits.plist
sudo shutdown -r now
```
And the procedure for disabling SIP is [here](https://developer.apple.com/library/content/documentation/Security/Conceptual/System_Integrity_Protection_Guide/ConfiguringSystemIntegrityProtection/ConfiguringSystemIntegrityProtection.html). Note: you'll have to replace `enable` with `disable` in the command given on that page. 

Once all of this is done, you should absolutely re-enable SIP, as leaving it off is less secure.

If anyone wants to volunteer let me know if you want me to screenshare with you and we'll schedule something. Otherwise let me know how you plan on providing us with the core file.

**Comment by NDuggan** [comment]
@CamJN sorry but I'd have to opt-out of this one. As my MBP belongs to my employer and they have already told us not to disable SIP.

**Comment by wulffeld** [comment]
@CamJN I've just done what you asked and the app crashes but nothing appears in /cores. Am I looking in the wrong place?

**Comment by CamJN** [comment]
@wulffeld `/cores` is the correct location. If you run `launchctl limit core` does it say `core        unlimited      unlimited`? And what does `csrutil status` output?

**Comment by wulffeld** [comment]
@CamJN ```# launchctl limit core
	core        unlimited      unlimited      
```

**Comment by wulffeld** [comment]
@CamJN ```# csrutil status
System Integrity Protection status: disabled.
```

**Comment by CamJN** [comment]
And that doesn't produce a core when Apache crashes. That is truly bizarre, as that's the criteria for generating a coredump of a system process these days.

Oh well, since you've disabled SIP, you can manually attach lldb to apache, create a breakpoint at the relevant function and then save the core file. The process for that is as follows:

```
lldb -p $(ps -A | fgrep -i httpd | sort -n | head -1 | awk '{print $1}')
b objc_initializeAfterForkError
process continue
#once the process stops on the above breakpoint do the following:
process save-core /cores/httpd.core
```

**Comment by wulffeld** [comment]
@CamJN Okay, so I think at first it attached to the wrong httpd but I got it attached to the right one and:

```
# ls -l /cores/
total 2393496
-rw-------  1 root  admin  1211322368 Nov 20 21:20 httpd.core
```

Does that seem like the correct size for a core file? 😬

**Comment by CamJN** [comment]
Yeah they're huge. 1.13 GB is not surprising. So now you can either send us the file, or if you prefer, we can setup a time to screen-share over Skype. I believe Apple's Maildrop supports files up to 5GB, or if you have a preferred method of transferring huge files that's cool too. (Maybe post a hash so I can verify the file wasn't corrupted in transit.)

**Comment by wulffeld** [comment]
@CamJN Email sent with info.

**Comment by CamJN** [comment]
Well that made things much clearer! Can everyone confirm that their systems are running case-insensitive filesystems (which is the default for either HFS+ or APFS)?

**Comment by NDuggan** [comment]
@CamJN I can confirm I am running a case-insensitive filesystem.

```
$ diskutil info /
File System Personality:  APFS
Type (Bundle):            apfs
Name (User Visible):      APFS
```

![image](https://user-images.githubusercontent.com/3830636/33063101-5a7b4dea-ce99-11e7-87d4-14008d93d7f0.png)

**Comment by wulffeld** [comment]
@CamJN case insensitive here as well on Journaled HFS+.

**Comment by CamJN** [comment]
Ok that confirms my suspicion. Apple introduced a bug into the system Apache in High Sierra, you can review the change yourself if it interests you here: 

The original file:
https://opensource.apple.com/source/apache/apache-822/httpd/server/request.c.orig.auto.html
Apple's modified file:
https://opensource.apple.com/source/apache/apache-822/httpd/server/request.c.auto.html

Particularly, they call `CFStringLowercase` in `files_strcmp`, which after a long call tree loads the Obj-C runtime (something Core Foundation is not supposed to do). 

What's interesting to me is that when an older Passenger is loaded this does not occur, and presumably this also doesn't occur when `mod_passenger` is not loaded at (maybe someone could check that case). Meaning that this `files_strcmp` function is triggered because `mod_passenger` is loaded.

I'll poke around to see if I can find a way to avoid triggering the bug.

**Comment by wulffeld** [comment]
@CamJN Not sure how to check it by not loading mod_passenger?

My 5.1.2 was compiled on Sierra, I'm quite sure - in case that matters. I tried compiling 5.1.2 on High Sierra but can't because of the include file changes IIRC.

**Comment by CamJN** [comment]
@wulffeld just remove/comment-out the passenger related configs from your apache config (or wrap them in `<IfModule mod_passenger.c>…</IfModule>`) along with the `LoadModule passenger_module /usr/local/opt/passenger/libexec/buildout/apache2/mod_passenger.so` line.

**Comment by wulffeld** [comment]
@CamJN Yeah, apache works fine without the module as well.

**Comment by CamJN** [comment]
@wulffeld can you try adding `PassengerHighPerformance on` to an affected virtualhost? I think that might bypass the problematic function.

**Comment by wulffeld** [comment]
@CamJN That makes it work. 👍

**Comment by wulffeld** [comment]
@CamJN Well... the Rails app loads but references to images (jpg, png, etc.) don't work anymore. By nature of it bypassing certain modules I assume from reading the docs.

**Comment by CamJN** [comment]
Well at least that narrows it down. (Also some people will be able to work around this bug that way).

**Comment by CamJN** [comment]
So the current workarounds are:

- Use a non-system apache (so it doesn't have apple's patches)
- Set the `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` env var to the apache launchd plist (requires disabling SIP)
- Add `PassengerHighPerformance on` to your virtualhost if your app can work that way
- Use a case-sensitive filesystem (requires a fresh install of macOS)

**Comment by patricksereno** [comment]
On High Sierra (10.13) with X-Code 9.1, I was able to install the ruby gem (5.1.12), along with the patch to make it compile:

    *** apache-fix.rb	2017-12-11 23:26:43.000000000 -0700
    --- apache.rb	2017-12-11 23:26:21.000000000 -0700
    *************** module PhusionPassenger
    *** 614,619 ****
    --- 614,620 ----
              if os_name_simple == 'macosx' && os_version >= '10.13' && httpd == '/usr/sbin/httpd'
                xcode_prefix = `/usr/bin/xcode-select -p`.strip
                flags << "-I#{xcode_prefix}/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/apache2"
    +           flags << "-I/usr/include/apache2"
              end
            else
              apxs2_flags = `#{apxs2} -q CFLAGS`.strip << " -I" << `#{apxs2} -q INCLUDEDIR`.strip
    *************** module PhusionPassenger
    *** 816,821 ****
    --- 817,823 ----
                  xcode_prefix = `/usr/bin/xcode-select -p`.strip
                  ["-I#{xcode_prefix}/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/apr-1",
                '-lapr-1']
    +             ['-I/usr/include/apr-1', '-lapr-1']
                else
                  ['-I/usr/include/apr-1', '-lapr-1']
                end

And then include in the VirtualHost configuration:

`PassengerHighPerformance on`

However, trying to access a file from public results in the following apache error message:

    objc[20681]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
    objc[20681]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
    [Mon Dec 11 22:36:53.335695 2017] [core:notice] [pid 20493] AH00052: child pid 20681 exit signal Abort trap (6)

I would be inclined to blame the Sinatra 2.0.0 gem, but this is the same error that was occurring for '/' before.

(Of course, this is much better than the previous state.)

**Comment by CamJN** [comment]
Yeah, it's going to be a game of whack a mole until Apple fixes their end. You can help that happen faster by opening bug reports at https://bugreport.apple.com/ as they prioritize duplicated bugs.

**Comment by mattbob** [comment]
Any update on this? I tried what @NDuggan suggested using the brew version of apache but that didn't fix the issue for me. `passenger-status` and `passenger-config validate-install` all look good but I get this error over and over
```
objc[61064]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[61064]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[Mon Jan 01 16:44:20.814012 2018] [core:notice] [pid 61044] AH00052: child pid 61064 exit signal Abort trap (6)
```

Passenger: 5.1.12 - Apache 2.4.28 - macOS: 10.13.2

**Comment by CamJN** [comment]
@mattbob if there's a crash in the homebrew apache, that would be surprising, as it doesn't have the patches that invoke the Obj-C runtime to load. Can you get a backtrace?

**Comment by enrique-ramirez** [comment]
I'm not sure if this is the same error as @mattbob .

I'm running a ruby app on an SSL virtual host using passenger. This is my config: https://gist.github.com/enrique-ramirez/6173de71b866ce2e3b6cb78a94b84fe7

I also want to be able to access `localhost` via http (port 80) and `localhost/~userName` folders.

For some reason, I cannot get both of them running. The gist config works fine for my ruby app, but when attempting to access `localhost` (default Apache DocumentRoot), it always says `ERR_EMPTY_RESPONSE` and when checking the logs, I see the exact same error @mattbob has:

```
objc[27567]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[27567]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
[Thu Jan 04 16:41:04.006967 2018] [core:notice] [pid 27258] AH00052: child pid 27567 exit signal Abort trap (6)
```

If I do not load the `passenger_module`, apache is back to normal but, of course, my rails app does not run.

I'm sorry if this is already answered, but there are so many comments it's a bit confusing: Is there a way to get both of them running? I develop a few PHP projects, so I need apache to run them as well.

**Comment by CamJN** [comment]
Yes, that sounds like the same bug, please file a bug report with apple, the more they get the sooner they'll fix this.

**Comment by vinz84** [comment]
@enrique-ramirez deactivate apache that come with osx and reinstall like this instruction:
https://getgrav.org/blog/macos-sierra-apache-multiple-php-versions

Will work for me (passenger 5.1.12, Xcode 9.1, High Sierra 10.13.2, ruby-2.3.1 via rvm)

**Comment by enrique-ramirez** [comment]
@vinz84 Thanks! I'll give it a go once I have a bit more time. I resorted to MAMP for the time being.

**Comment by CamJN** [comment]
Closing this issue, because the original problem was found in Apple's Apache. If you have a problem that can be shown to not involve the system Apache then you can open another issue.
